amulet-map-editor: init at 0.10.44

Tibso opened the pull request:
This PR adds the Amulet Map Editor package to Nixpkgs.
This application is a map editor for the video game Minecraft.

The application has `python-modules` dependencies that I also added individually, in case people want to use them for some Minecraft Python projects.

This closes [#201016](https://github.com/NixOS/nixpkgs/issues/201016)

And I suppose also closes [#356035](https://github.com/NixOS/nixpkgs/pull/356035)
Some peeps on Discord told me there was already a PR for this package, but silly me must have forgotten to check for a previous PR. So I've quickly added some stuff from the other PR (thank you @pluiedev) but I did not include stuff like tests and Sphinx docs stuff because I haven't looked into those yet, so I don't know how those things work.
Tell me if I need to add those for this PR.

The application works and is usable, although there are some UI rendering issues related to `wxpython`, which make the application hard to use. The application devs are aware of the UI issues (seems to be related to this issue [#127](https://github.com/Amulet-Team/Amulet-Map-Editor/issues/127)) and they seem to occur for any linux user (so also for other distros).
The devs are working on replacing `wxpython` to fix the UI issues.

## Things done

- Added myself to `maintainers.list`
- Added `amulet` to `licenses.list`
- Added the `amulet-map-editor` application

The following `python-modules` were also added:
- amulet-core
- amulet-leveldb
- amulet-nbt
- minecraft-resource-pack
- pymctranslate

Building and testing:

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Tibso committed:
maintainers: add Tibso

Tibso committed:
python3Packages.amulet-nbt: init at 2.1.5

Co-authored-by: Leah Amelia Chen <hi@pluie.me>
Co-authored-by: langsjo <104687438+langsjo@users.noreply.github.com>

Tibso committed:
python3Packages.amulet-leveldb: init at 1.0.2

Co-authored-by: Leah Amelia Chen <hi@pluie.me>

Tibso committed:
python3Packages.pymctranslate: init at 1.2.35

Co-authored-by: Leah Amelia Chen <hi@pluie.me>

Tibso committed:
python3Packages.amulet-core: init at 1.9.30

Co-authored-by: Leah Amelia Chen <hi@pluie.me>

Tibso committed:
python3Packages.minecraft-resource-pack: init at 1.4.6

Co-authored-by: Leah Amelia Chen <hi@pluie.me>

Tibso committed:
amulet-map-editor: init at 0.10.44

Co-authored-by: Leah Amelia Chen <hi@pluie.me>

Sigmanificient reviewed (commented):
Hi, welcome to nixpkgs and thanks you for your awesome work!

Theses two change request apply to all the derivation in order to follow current `buildPython{Application,Package}` standards, sorry this is quite some changes to do.
Lets also make sure we can importa theses packages using `pyrhonImportsCheck`!

pluiedev commented:
Hey :) I'm okay with closing my PR in favor of yours if I could be credited as a co-author on changes that were taken from my PR (add `Co-authored-by: Leah Amelia Chen <hi@pluie.me>` at the very bottom of the commit message). No need to add me as a maintainer though, as I was only trying to package this as an experiment and out of vested personal interest.

FliegendeWurst commented on pkgs/development/python-modules/amulet-core/default.nix:
```diff
@@ -0,0 +1,67 @@
+    amulet-leveldb
+    pymctranslate
+    lz4
+    (portalocker.overridePythonAttrs (_: {
```
Why are these overriden to specific versions?

Tibso replied:
These dependencies are too recent in Nixpkgs, so I need to override them with older versions.

FliegendeWurst replied:
>  This is to meet the requirements.py, otherwise the runtime dependencies check fails.

I think we ignore such strict requirements in nixpkgs, unless they are actually required. For example, `portalocker` 3.0 (despite the major version upgrade) shouldn't have any breaking changes relevant to amulet-core.

Tibso replied:
So the modules and the application would still work properly even if the runtime dependencies check fails?

FliegendeWurst replied:
Yes, of course you would need to check that the application works properly / review the breaking changes in the libraries.

Sigmanificient commented on pkgs/by-name/am/amulet-map-editor/default.nix:
```diff
@@ -0,0 +1,57 @@
+  };
+
+  buildInputs = with python3.pkgs; [
+    setuptools
```
Yes, you want all your python modules to use `pyproject = true` so they will also uses `build-system`, `dependencies` and most importantly the `pythonImportsCheck`

FliegendeWurst commented on pkgs/development/python-modules/amulet-core/default.nix:
```diff
@@ -0,0 +1,67 @@
+  meta = {
+    description = "Python library for reading and writing the Minecraft save formats";
+    homepage = "https://www.amuletmc.com/";
+    license = with lib.licenses; [ amulet ];
```
This repository doesn't have a license at all (https://github.com/Amulet-Team/Amulet-Core/issues/155)

Tibso replied:
The main dev told me to add their team license to all modules and that I could point to the license on the [Amulet-NBT](https://github.com/Amulet-Team/Amulet-NBT) repo. But I agree that this isn't clean.

FliegendeWurst replied:
Then it's alright I guess.. could you change the `meta.homepage` of the libraries to point to their Github page?

FliegendeWurst replied:
And can you add `nativeCheckInputs = [ pytestCheckHook ]` (if they use pytest)? I see that some repos have a test suite.

FliegendeWurst replied:
It would be good to have tests passing, especially if we relax the dependency version requirements.

Tibso commented:
I've fixed all mentioned issues.
Let me know if there's still something to change or add.

Tibso commented:
I've fixed mentioned issues again, minus the modules tests because they fail for reasons I don't understand.
I'll look into it if it is necessary for this PR.

alyssais reviewed (commented):
Ack license.

Tibso commented:
Okay now we have the tests working and passing except on `minecraft-resource-pack`. I disabled those tests because the dev told me they were broken for that one module.

I asked for help on our Discord server to add the tests. @langsjo discovered that the reason I was having issues was because the modules I fetched from PyPI didn’t include any tests. The solution was to use `fetchFromGitHub` instead of `fetchPypi` to get the modules with the test files included... So basically like on the older PR...

I sincerely hope this PR is ready to go now.
Let me know. = )

emilazy commented:
If I read that PR correctly, there was no licence in the repository at *all* previously, and they have just pinged every past contributor saying they’re going to assume their changes were under MIT without there having previously been any indication of that?

That is, uh, not how it works. If I understand correctly, then this is going to have to be `lib.licenses.unfree` with a big warning in a comment forever unless all the previous contributors agree to the change or their contributions are rewritten or can be shown to be too trivial to be copyrightable.

pluiedev commented:
As I understand it the core maintainer(s) weren't really concerned about the legal implications of unclear licenses for us packagers. It feels like an unspoken agreement that the repositories *should be* in their custom Polyform Shield fork, and they never bothered to codify it as such.

I personally wouldn't take any chances and would only label the packages with an explicit LICENSE file with `lib.licenses.amulet`, and  the others with unfree.

emilazy commented:
> I guess it really is time for [`lib.licenses.illegal`](https://github.com/NixOS/nixpkgs/issues/404663#issuecomment-2855478207) to be a thing then 😅

Oh, we’ve been there for a long time…

`lib.licenses.unfree` is how we tend to handle this kind of thing. I personally don’t love adding new packages with messed up copyright status but there is a lot of questionable stuff already in the tree.

> Looks like they started using `CLAassistant` since [Amulet-Team/Amulet-Map-Editor#530](https://github.com/Amulet-Team/Amulet-Map-Editor/pull/530) though (May 2022), so there exist some changes prior that are not "MIT-able", most notably [ALTB0ne](https://github.com/Amulet-Team/Amulet-Map-Editor/commits?author=ALTB0ne)'s (translations) and [naor2013](https://github.com/Amulet-Team/Amulet-Map-Editor/commits?author=naor2013)'s (a couple minor features) ...

These are clearly substantial enough contributions to be copyrightable. If they don’t have permission from these contributors then we can’t record this as Polyform regardless of what the repository says.

Tibso commented:
Okay, so if I understand the licensing issue correctly, the Amulet license is somehow flawed.
Therefore, I have to remove the Amulet license (`lib.licenses.amulet`) and simply use `lib.licenses.unfree` for the Amulet application and the python modules?
Or the scuffed Amulet license is fine as it is configured now?

I looked at #422452 and there seem to be other issues I haven't encountered myself that would need fixing, but I could work with @erkingoksuu on a following PR to add his fixes.

I'll update this PR for version 0.10.44 whenever I have a bit of time.

I'd really like to be done with this PR as it adds me in the maintainer list and there are other PRs that I would like to do. =)

dtomvan commented:
I think you'd need to carefully look which project do and don't include a LICENSE file with the amulet license in it. If it doesn't it should be changed to `lib.licenses.unfree`, since we cannot truthfully call it "licensed under the amulet license".

For example, `amulet-nbt` is correct with `lib.licenses.amulet`, as it [does have the amulet license in its tree](https://github.com/Amulet-Team/Amulet-NBT/blob/5.0/LICENSE)

Tibso commented:
Okay I updated the editor and the modules and changed the metadatas to use `lib.licenses.unfree` whenever the Amulet license was not present in their respective repository.

I retested the editor and it still works with the UI issues already mentioned before.There are also random segfaults apparently caused by Pango that I forgot to mention before. But its apparently another issue that users experience (https://github.com/Amulet-Team/Amulet-Map-Editor/issues/1012).

UlyssesZh commented:
I think there is no point in defining `lib.licenses.amulet` when even PolyForm licenses are not in `lib.licenses`. What current packages in nixpkgs do is just put unfree and add a comment with the link to the upstream license file.

Tibso commented:
Can someone else confirm that once I do the following:

> What current packages in nixpkgs do is just put unfree and add a comment with the link to the upstream license file.

... this PR will be merged? Because it's been a while now...
